File: Payload_Type/poseidon/mythic/agent_functions/unlink_tcp.py
from mythic_payloadtype_container.MythicCommandBase import *
import json


class UnLinkTCPArguments(TaskArguments):
    def __init__(self, command_line, **kwargs):
        super().__init__(command_line, **kwargs)
        self.args = [
            CommandParameter(
                name="connection",
                type=ParameterType.LinkInfo,
                description="Connection info for unlinking",
                parameter_group_info=[
                    ParameterGroupInfo(ui_position=1)
                ]
            ),
        ]

    async def parse_arguments(self):
        if len(self.command_line) == 0:
            raise Exception("Must provide arguments")
        else:
            try:
                self.load_args_from_json_string(self.command_line)
            except:
                raise Exception("Failed to load arguments as JSON. Did you use the popup?")

    async def parse_dictionary(self, dictionary):
        self.load_args_from_dictionary(dictionary)


class UnLinkTCPCommand(CommandBase):
    cmd = "unlink_tcp"
    needs_admin = False
    help_cmd = "unlink_tcp"
    description = "Unlink a tcp connection."
    version = 1
    author = "@its_a_feature_"
    argument_class = UnLinkTCPArguments
    attackmapping = []

    async def create_tasking(self, task: MythicTask) -> MythicTask:
        connection_info = task.args.get_arg("connection")
        remote_agent_id = connection_info.pop("callback_uuid", None)
        if remote_agent_id is None or remote_agent_id == "":
            raise Exception("Missing callback UUID in connection information")
        task.args.remove_arg("connection")
        task.args.add_arg("connection", remote_agent_id, type=ParameterType.String)
        task.display_params = f"from {remote_agent_id}"
        return task

    async def process_response(self, response: AgentResponse):
        pass

File: Payload_Type/poseidon/poseidon/agentfunctions/unlink_tcp.go
package agentfunctions

import (
	"fmt"
	agentstructs "github.com/MythicMeta/MythicContainer/agent_structs"
)

func init() {
	agentstructs.AllPayloadData.Get("poseidon").AddCommand(agentstructs.Command{
		Name:                "unlink_tcp",
		Description:         "Unlink a poseidon_tcp connection.",
		HelpString:          "unlink_tcp",
		Version:             1,
		MitreAttackMappings: []string{},
		SupportedUIFeatures: []string{},
		Author:              "@its_a_feature_",
		CommandAttributes: agentstructs.CommandAttribute{
			SupportedOS: []string{},
		},
		CommandParameters: []agentstructs.CommandParameter{
			{
				Name:          "connection",
				Description:   "Connection info for unlinking",
				ParameterType: agentstructs.COMMAND_PARAMETER_TYPE_LINK_INFO,
			},
		},
		TaskFunctionParseArgString: func(args *agentstructs.PTTaskMessageArgsData, input string) error {
			return nil
		},
		TaskFunctionParseArgDictionary: func(args *agentstructs.PTTaskMessageArgsData, input map[string]interface{}) error {
			return args.LoadArgsFromDictionary(input)
		},
		TaskFunctionCreateTasking: func(taskData *agentstructs.PTTaskMessageAllData) agentstructs.PTTaskCreateTaskingMessageResponse {
			response := agentstructs.PTTaskCreateTaskingMessageResponse{
				Success: true,
				TaskID:  taskData.Task.ID,
			}
			if connectionInfo, err := taskData.Args.GetLinkInfoArg("connection"); err != nil {
				response.Success = false
				response.Error = err.Error()
			} else if connectionInfo.CallbackUUID == "" {
				response.Success = false
				response.Error = "Failed to find callback UUID in connection information"
			} else {
				taskData.Args.RemoveArg("connection")
				taskData.Args.AddArg(agentstructs.CommandParameter{
					Name:          "connection",
					DefaultValue:  connectionInfo.CallbackUUID,
					ParameterType: agentstructs.COMMAND_PARAMETER_TYPE_STRING,
				})
				displayString := fmt.Sprintf("from %s", connectionInfo.CallbackUUID)
				response.DisplayParams = &displayString
			}
			return response
		},
	})
}
